sql: add substring_index built-in function

spilchen opened the pull request:
This commit introduces the `substring_index` built-in function, mirroring MySQL's behavior. `substring_index` returns a substring of `input` before `count` occurrences of `delim`. If `count` is positive, the leftmost part is returned. If `count` is negative, the rightmost part is returned.

```
demo@127.0.0.1:26257/demoapp/db> SELECT substring_index('www.cockroachlabs.com', '.', 2);
   substring_index
---------------------
  www.cockroachlabs
(1 row)

Time: 3ms total (execution 3ms / network 0ms)

demo@127.0.0.1:26257/demoapp/db> SELECT substring_index('www.cockroachlabs.com', '.', -2);
   substring_index
---------------------
  cockroachlabs.com
(1 row)

Time: 1ms total (execution 1ms / network 0ms)
```

[TREQ-900](https://cockroachlabs.atlassian.net/browse/TREQ-900)
Epic: none
Release note (sql change): Added the `substring_index` built-in function, which extracts a portion of a string based on a specified delimiter and occurrence count, following MySQL behavior.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141929)
<!-- Reviewable:end -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141929)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/sem/builtins/builtins.go` line 400 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141929#-OJt4Kk_4q3mSVTkgkJ9:-OJt4Kk_4q3mSVTkgkJA:bipz4li) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec4db4cdc1f65e1c159f1e02b16aa9d7a1831128/pkg/sql/sem/builtins/builtins.go#L400)):*
> ```Go
> 			ReturnType: tree.FixedReturnType(types.String),
> 			Fn: func(_ context.Context, _ *eval.Context, args tree.Datums) (tree.Datum, error) {
> 				// Handle NULL behaviour
> ```

for function overloads, there's a field called `CalledOnNullInput`. for most functions, the desired behavior is that if any of the arguments is null, the output is null, so when `CalledOnNullInput` is false (which is the default), that is what happens.

but if there are any cases in which a null argument has a non-null result, we should make sure to set `CalledOnNullInput` to true.

in this case, it looks like `CalledOnNullInput=false` is desired.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141929#-:-OJt6Lv9EI4lk5HIPVQp:b-j4xmku)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)



<!-- Sent from Reviewable.io -->

spilchen committed:
sql: add substring_index built-in function

This commit introduces the `substring_index` built-in function,
mirroring MySQL's behavior. `substring_index` returns a substring of
`input` before `count` occurrences of `delim`. If `count` is positive,
the leftmost part is returned. If `count` is negative, the rightmost
part is returned.

Epic: none
Release note (sql change): Added the `substring_index` built-in
function, which extracts a portion of a string based on a specified
delimiter and occurrence count, following MySQL behavior.

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141929)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/sem/builtins/builtins.go` line 400 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141929#-OJt4Kk_4q3mSVTkgkJ9:-OJtJ6p45ypDGgG9P1Eo:b-30pztv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec4db4cdc1f65e1c159f1e02b16aa9d7a1831128/pkg/sql/sem/builtins/builtins.go#L400)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

for function overloads, there's a field called `CalledOnNullInput`. for most functions, the desired behavior is that if any of the arguments is null, the output is null, so when `CalledOnNullInput` is false (which is the default), that is what happens.

but if there are any cases in which a null argument has a non-null result, we should make sure to set `CalledOnNullInput` to true.

in this case, it looks like `CalledOnNullInput=false` is desired.
</blockquote></details>

This is good to know. I confirmed that NULLs are correctly handled based on the default `CalledOnNullInput` value, so I removed the additional NULL handling I had added.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
